pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java

```diff
@@ -120,6 +126,49 @@ public boolean hasNullValues() {
     return !_nullValueFields.isEmpty();
   }
 
+  /**
+   * @return a deep copy of the generic row
+   */
+  public GenericRow copy() {
+    GenericRow copy = new GenericRow();
+    copy.init(this);
+    for (Map.Entry<String, Object> entry : new HashSet<>(copy._fieldToValueMap.entrySet())) {
+      copy.putValue(entry.getKey(), copy(entry.getValue()));
+    }
```

Comment:
```suggestion
    for (Map.Entry<String, Object> entry : copy._fieldToValueMap.entrySet()) {
      entry.setValue(copy(entry.getValue()));
    }
```

Reply:
this is no longer a deep copy?

Reply:
This is still deep copy as we switched the value to a deep copy one